xds/pkg/server/server.go

```diff
@@ -33,40 +38,101 @@ const (
 	grpcMaxConcurrentStreams = 1000000
 )
 
+// HealthCheck implements the health check logic exposed by the server.
+// It is implemented by internal components of the server and returns
+// an error if the component is unhealthy.
+type HealthCheck func(ctx context.Context) error
+
 // Server is a wrapper around go-control-plane's xds server.
 type Server struct {
 	logger *log.Logger
-	// port is the port the server runs on.
-	port int16
+	// xdsPort is the port the XDS server runs on.
+	xdsPort int16
 	// snapshotCache is passed updated by us and read by go-control-plane
 	// to provided proxies with config from.
 	snapshotCache cache.SnapshotCache
 	// nodeIDCh is nilable. It is used as a callback mechanism from go-control-plane's server to
 	// let us know when a new proxy has connected. The proxy's ID is passed on the channel.
 	nodeIDCh chan<- string
+	// healthChecks contains the server's health check implementations.
+	healthChecks []HealthCheck
+	// adminPort is the port the admin server runs on.
+	adminPort int16
 }
 
-// NewServer returns a new Server
+// New returns a new Server
 func New(
 	logger *log.Logger,
-	port int16,
+	xdsPort int16,
 	snapshotCache cache.SnapshotCache,
 	nodeIDCh chan<- string,
+	healthChecks []HealthCheck,
+	adminPort int16,
 ) *Server {
 	return &Server{
 		logger: logger.WithFields(log.Fields{
 			"component": "server",
 		}).Logger,
-		port:          port,
+		xdsPort:       xdsPort,
 		snapshotCache: snapshotCache,
 		nodeIDCh:      nodeIDCh,
+		healthChecks:  healthChecks,
+		adminPort:     adminPort,
 	}
 }
 
+func (s *Server) startAdminServer(ctx context.Context) {
+	address := fmt.Sprintf(":%d", s.adminPort)
+
+	srv := &http.Server{Addr: address}
+
+	http.Handle("/metrics", promhttp.Handler())
+
+	http.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
```

Comment:
Is this meant to be a liveness probe or a readiness probe?

It would be good if this was `/live` or `/ready` so that it was clear.

I'm assuming that it's a readiness probe, in that if the k8s control plane goes down, we don't want to keep restarting this process because the health check is failing.

In that vein, it's probably useful to have both, as if the process can't respond to a simple http response, it should probably be restarted.

But if the k8s api is down we should keep logging this via liveness, but not restart the process.

Reply:
Updated to /ready, I assume if folks want to use it for liveness they can point to the same endpoint as well?

Reply:
Yeah they could. But I don't think people will want to restart the container if the K8s control plane goes down (but it's a choice 😁)

Doesn't have to be this PR, but I'd be tempted to have a `/live` endpoint as well that was just a Http Handler that returns 200. 

Reason being that if the process does freeze (and can't respond to just a simple http request) for whatever reason, then it will restart and hopefully recover again.